app/src/chat_engine.py

```diff
@@ -0,0 +1,146 @@
+import logging
+from abc import ABC, abstractmethod
+from typing import Type
+
+import chainlit as cl
+import src.adapters.db as db
+from chainlit.input_widget import Slider, Switch
+from src.db.models.document import Chunk
+from src.format import format_guru_cards
+from src.generate import generate
+from src.retrieve import retrieve
+from src.shared import get_embedding_model
+
+logger = logging.getLogger(__name__)
+
+
+class ChatEngineInterface(ABC):
+    engine_id: str
+    name: str
+
+    @abstractmethod
+    async def on_start(self) -> dict:
+        pass
+
+    @abstractmethod
+    def on_message(self, question: str, cl_message: cl.Message) -> dict:
+        pass
+
+    @abstractmethod
+    def format_answer_message(self, results: dict) -> str:
+        pass
+
+
+def all_subclasses(cls: Type) -> set:
```

Comment:
Bit messy but I think you should be able to do something like:
```
from typing import Type, TypeVar

T = TypeVar('T')

def all_subclasses(cls: Type[T]) -> set[Type[T]]:

Reply:
(and a sidenote that this is a really slick way of automatically keeping the list of available engines up to date 👍 )

Reply:
https://github.com/navapbc/labs-decision-support-tool/actions/runs/9958216005/job/27512064988?pr=25#step:4:16

Reply:
Ah, unfortunately it looks like my suggestion doesn't work for abstract classes. Yet another reason to wish for a better typing system for Python 😭